staging/src/k8s.io/pod-security-admission/cmd/webhook/app/server.go

```diff
@@ -0,0 +1,261 @@
+}
+
+func (s *Server) Start(ctx context.Context) error {
+	s.informerFactory.Start(ctx.Done())
+
+	mux := http.NewServeMux()
+	healthz.InstallHandler(mux, healthz.PingHealthz)
+	healthz.InstallReadyzHandler(mux, healthz.NewInformerSyncHealthz(s.informerFactory))
+	// The webhook is stateless, so it's safe to expose everything on the insecure port for
+	// debugging or proxy purposes. The API server will not connect to an http webhook.
+	mux.HandleFunc("/", s.HandleValidate)
+
+	if s.insecureServing != nil {
+		if err := s.insecureServing.Serve(mux, 0, ctx.Done()); err != nil {
+			return fmt.Errorf("failed to start insecure server: %w", err)
+		}
+	}
+
+	var shutdownCh <-chan struct{}
+	if s.secureServing != nil {
+		var err error
+		shutdownCh, err = s.secureServing.Serve(mux, 0, ctx.Done())
+		if err != nil {
+			return fmt.Errorf("failed to start secure server: %w", err)
+		}
+	}
+
+	// Wait for graceful shutdown.
+	<-shutdownCh
+
+	return nil
+}
+
+func (s *Server) HandleValidate(w http.ResponseWriter, r *http.Request) {
+	defer utilruntime.HandleCrash(func(_ interface{}) {
+		// Assume the crash happened before the response was written.
+		writeAdmissionResponse(w, admission.InternalErrorResponse("internal server error"))
+	})
+
+	var (
+		body []byte
+		err  error
+		ctx  = r.Context()
+	)
+
+	if r.Body == nil {
+		err = errors.New("request body is empty")
+		klog.ErrorS(err, "bad request")
+		writeAdmissionResponse(w, admission.BadRequestResponse("bad request"))
+		return
+	}
+
+	defer r.Body.Close()
+	if body, err = ioutil.ReadAll(r.Body); err != nil {
+		klog.ErrorS(err, "unable to read the body from the incoming request")
+		writeAdmissionResponse(w, admission.BadRequestResponse("unable to read the body from the incoming request"))
+		return
+	}
+
+	// verify the content type is accurate
+	if contentType := r.Header.Get("Content-Type"); contentType != "application/json" {
+		err = fmt.Errorf("contentType=%s, expected application/json", contentType)
+		klog.ErrorS(err, "unable to process a request with an unknown content type", "content type", contentType)
+		writeAdmissionResponse(w, admission.BadRequestResponse("unable to process a request with a non-json content type"))
+		return
+	}
+
+	// Both v1 and v1beta1 AdmissionReview types are exactly the same, so the v1beta1 type can
+	// be decoded into the v1 type. However the runtime codec's decoder guesses which type to
+	// decode into by type name if an Object's TypeMeta isn't set. By setting TypeMeta of an
+	// unregistered type to the v1 GVK, the decoder will coerce a v1beta1 AdmissionReview to v1.
+	// The actual AdmissionReview GVK will be used to write a typed response in case the
+	// webhook config permits multiple versions, otherwise this response will fail.
+	review := &admissionv1.AdmissionReview{}
+	review.SetGroupVersionKind(admissionv1.SchemeGroupVersion.WithKind("AdmissionReview"))
+	if _, _, err := codecs.UniversalDeserializer().Decode(body, nil, review); err != nil {
```

Comment:
don't ignore the return object value... you're not guaranteed it is the same as the object you passed in

I expected something closer to:

```
review, gvk, err := codecs.UniversalDeserializer().Decode(body, nil, &admissionv1.AdmissionReview{})
if err != nil {
  ...
}
if gvk != v1AdmissionReview {
  ...
}
...
```